File: Tests/DraftRightTests/RewriteCacheKeyParityTests.swift
import XCTest
@testable import DraftRight

/// Cross-platform parity for the RewriteCache key. The key algorithm is
/// duplicated in three languages (macOS Swift, Windows C#, Linux Python); this
/// asserts the Swift `RewriteCache.key` reproduces every vector in the shared
/// `parity/rewrite-cache-key-vectors.json` fixture — the single source of truth
/// all three platforms check against (issue #174, guarding #147/#108). If this
/// fails, the Swift key format drifted from the others: fix the code, not the
/// fixture (unless the format was changed on purpose in all three).
final class RewriteCacheKeyParityTests: XCTestCase {

    private struct Vector: Decodable {
        let text: String
        let tone: String
        let language: String?
        let expectedKey: String
    }

    private struct Fixture: Decodable {
        let vectors: [Vector]
    }

    /// Locate the repo-root `parity/` fixture by walking up from this source
    /// file — robust to the SPM build layout and to running from CI or a dev box.
    private func fixtureURL() throws -> URL {
        var dir = URL(fileURLWithPath: #filePath).deletingLastPathComponent()
        while dir.path != "/" {
            let candidate = dir.appendingPathComponent("parity/rewrite-cache-key-vectors.json")
            if FileManager.default.fileExists(atPath: candidate.path) { return candidate }
            dir = dir.deletingLastPathComponent()
        }
        throw XCTSkip("parity/rewrite-cache-key-vectors.json not found walking up from \(#filePath)")
    }

    func testKeyMatchesSharedGoldenVectors() throws {
        let data = try Data(contentsOf: try fixtureURL())
        let fixture = try JSONDecoder().decode(Fixture.self, from: data)
        XCTAssertFalse(fixture.vectors.isEmpty, "fixture must contain at least one vector")
        for v in fixture.vectors {
            XCTAssertEqual(
                RewriteCache.key(text: v.text, tone: v.tone, language: v.language),
                v.expectedKey,
                "key drift for text=\(v.text) tone=\(v.tone) language=\(String(describing: v.language))"
            )
        }
    }
}

File: DraftRightWindows/DraftRightWindows.PureTests/RewriteCacheKeyParityTests.cs
using System.Text.Json;
using DraftRightWindows.Services;
using Xunit;

namespace DraftRightWindows.PureTests;

/// <summary>
/// Cross-platform parity for the RewriteCache key. The key algorithm is
/// duplicated in three languages (macOS Swift, Windows C#, Linux Python);
/// this test asserts the C# <see cref="RewriteCache.Key"/> reproduces every
/// vector in the shared <c>parity/rewrite-cache-key-vectors.json</c> fixture,
/// which is the single source of truth all three platforms are checked against.
/// See issue #174 (and #147/#108). If this fails, the C# key format drifted
/// from the other platforms — fix the code, not the fixture (unless the format
/// was changed on purpose in all three).
/// </summary>
public class RewriteCacheKeyParityTests
{
    public sealed record Vector(string Text, string Tone, string? Language, string ExpectedKey);

    private sealed record Fixture(Vector[] Vectors);

    /// <summary>
    /// Locate the repo-root <c>parity/</c> fixture by walking up from the test
    /// assembly's output directory — robust to the bin/Debug/... nesting and to
    /// running from CI, a dev box, or the IDE.
    /// </summary>
    private static string FixturePath()
    {
        var dir = new DirectoryInfo(AppContext.BaseDirectory);
        while (dir is not null)
        {
            var candidate = Path.Combine(dir.FullName, "parity", "rewrite-cache-key-vectors.json");
            if (File.Exists(candidate))
            {
                return candidate;
            }
            dir = dir.Parent;
        }
        throw new FileNotFoundException(
            "parity/rewrite-cache-key-vectors.json not found walking up from " + AppContext.BaseDirectory);
    }

    public static IEnumerable<object[]> Vectors()
    {
        var json = File.ReadAllText(FixturePath());
        var options = new JsonSerializerOptions { PropertyNameCaseInsensitive = true };
        var fixture = JsonSerializer.Deserialize<Fixture>(json, options)!;
        foreach (var v in fixture.Vectors)
        {
            yield return new object[] { v };
        }
    }

    [Theory]
    [MemberData(nameof(Vectors))]
    public void Key_MatchesSharedGoldenVector(Vector v)
    {
        Assert.Equal(v.ExpectedKey, RewriteCache.Key(v.Text, v.Tone, v.Language));
    }
}
